cmd/s3-zip-handlers.go

```diff
@@ -0,0 +1,496 @@
+	}
+
+	zipObjInfo, err := objectAPI.GetObjectInfo(ctx, bucket, zipPath, ObjectOptions{})
+	if err != nil {
+		// Return empty listing
+		return ListObjectsV2Info{}, nil
+	}
+
+	var zipInfo []byte
+
+	if z, ok := zipObjInfo.UserDefined[archiveInfoMetadataKey]; ok {
+		zipInfo = []byte(z)
+	} else {
+		zipInfo, err = updateObjectMetadataWithZipInfo(ctx, objectAPI, bucket, zipPath)
+	}
+
+	if err != nil {
+		return ListObjectsV2Info{}, err
+	}
+
+	files, err := zipindex.DeserializeFiles(zipInfo)
+	if err != nil {
+		return ListObjectsV2Info{}, err
+	}
+
+	sort.Slice(files, func(i, j int) bool {
+		return files[i].Name < files[j].Name
+	})
+
+	var (
+		count           int
+		isTruncated     bool
+		nextToken       string
+		listObjectsInfo ListObjectsV2Info
+	)
+
+	// Always set this
+	listObjectsInfo.ContinuationToken = token
+
+	// Open and iterate through the files in the archive.
+	for _, file := range files {
+		objName := zipObjInfo.Name + slashSeparator + file.Name
+		if objName <= startAfter || objName <= token {
+			continue
+		}
+		if strings.HasPrefix(objName, prefix) {
+			if count == maxKeys {
+				isTruncated = true
+				break
+			}
+			if delimiter != "" {
+				i := strings.Index(objName[len(prefix):], delimiter)
+				if i >= 0 {
+					commonPrefix := objName[:len(prefix)+i+1]
+					if len(listObjectsInfo.Prefixes) == 0 || commonPrefix != listObjectsInfo.Prefixes[len(listObjectsInfo.Prefixes)-1] {
+						listObjectsInfo.Prefixes = append(listObjectsInfo.Prefixes, commonPrefix)
+						count++
+					}
+					goto next
+				}
+			}
+			listObjectsInfo.Objects = append(listObjectsInfo.Objects, ObjectInfo{
+				Bucket:  bucket,
+				Name:    objName,
+				Size:    int64(file.UncompressedSize64),
+				ModTime: zipObjInfo.ModTime,
+			})
+			count++
+		}
+	next:
+		nextToken = objName
+	}
+
+	if isTruncated {
+		listObjectsInfo.IsTruncated = true
+		listObjectsInfo.NextContinuationToken = nextToken
+	}
+
+	return listObjectsInfo, nil
+}
+
+// getFilesFromZIPObject reads a partial stream of a zip file to build the zipindex.Files index
+func getFilesListFromZIPObject(ctx context.Context, objectAPI ObjectLayer, bucket, object string) (zipindex.Files, ObjectInfo, error) {
+	var size = 1 << 20
+	var objSize int64
+	for {
+		rs := &HTTPRangeSpec{IsSuffixLength: true, Start: int64(-size)}
+		gr, err := objectAPI.GetObjectNInfo(ctx, bucket, object, rs, nil, readLock, ObjectOptions{})
+		if err != nil {
+			return nil, ObjectInfo{}, err
+		}
+		b, err := stdioutil.ReadAll(gr)
+		if err != nil {
+			gr.Close()
+			return nil, ObjectInfo{}, err
+		}
+		gr.Close()
+		if size > len(b) {
+			size = len(b)
+		}
+
+		// Calculate the object real size if encrypted
+		if _, ok := crypto.IsEncrypted(gr.ObjInfo.UserDefined); ok {
+			objSize, err = gr.ObjInfo.DecryptedSize()
+			if err != nil {
+				return nil, ObjectInfo{}, err
+			}
+		} else {
+			objSize = gr.ObjInfo.Size
+		}
+
+		files, err := zipindex.ReadDir(b[len(b)-size:], objSize, nil)
```

Comment:
I think probably we need an API that takes zipindex.ReadDirStream(gr, objSize, nil) - we can avoid 

> b, err := stdioutil.ReadAll(gr)